samza-test/src/test/java/org/apache/samza/test/operator/StreamApplicationIntegrationTestHarness.java

```diff
@@ -0,0 +1,245 @@
+        1,
+        60 * 1000L,
+        1024L * 1024L,
+        0,
+        0L,
+        5 * 1000L,
+        SecurityProtocol.PLAINTEXT,
+        null,
+        Option$.MODULE$.<Properties>apply(new Properties()),
+        new StringSerializer(),
+        new StringSerializer(),
+        Option$.MODULE$.<Properties>apply(new Properties()));
+
+    consumer = TestUtils.createNewConsumer(
+        bootstrapServers(),
+        "group",
+        "earliest",
+        4096L,
+        "org.apache.kafka.clients.consumer.RangeAssignor",
+        30000,
+        SecurityProtocol.PLAINTEXT,
+        Option$.MODULE$.<File>empty(),
+        Option$.MODULE$.<Properties>empty(),
+        Option$.MODULE$.<Properties>apply(consumerDeserializerProperties));
+  }
+
+  /**
+   * Creates a kafka topic with the provided name and the number of partitions
+   * @param topicName the name of the topic
+   * @param numPartitions the number of partitions in the topic
+   */
+  public void createTopic(String topicName, int numPartitions) {
+    TestUtils.createTopic(zkUtils(), topicName, numPartitions, 1, servers(), new Properties());
+  }
+
+  /**
+   * Produces a message to the provided topic partition.
+   * @param topicName the topic to produce messages to
+   * @param partitionId the topic partition to produce messages to
+   * @param key the key in the message
+   * @param val the value in the message
+   */
+  public void produceMessage(String topicName, int partitionId, String key, String val) {
+    producer.send(new ProducerRecord(topicName, partitionId, key, val));
+    producer.flush();
+  }
+
+  @Override
+  public int clusterSize() {
+    return Integer.parseInt(KafkaConfig.TOPIC_DEFAULT_REPLICATION_FACTOR());
+  }
+
+
+  /**
+   * Read messages from the provided list of topics until {@param threshold} messages have been read or until
+   * {@link #NUM_EMPTY_POLLS} polls return no messages.
+   *
+   * The default poll time out is determined by {@link #POLL_TIMEOUT_MS} and the number of empty polls are
+   * determined by {@link #NUM_EMPTY_POLLS}
+   *
+   * @param topics the list of topics to consume from
+   * @param threshold the number of messages to consume
+   * @return the list of {@link ConsumerRecord}s whose size can be atmost {@param threshold}
+   */
+  public List<ConsumerRecord<String, String>> getMessages(Collection<String> topics, int threshold) {
+    int emptyPollCount = 0;
+    List<ConsumerRecord<String, String>> recordList = new ArrayList<>();
+    consumer.subscribe(topics);
+
+    while (emptyPollCount < NUM_EMPTY_POLLS && recordList.size() < threshold) {
+      ConsumerRecords<String, String> records = consumer.poll(POLL_TIMEOUT_MS.toMillis());
+      if (!records.isEmpty()) {
+        Iterator<ConsumerRecord<String, String>> iterator = records.iterator();
+        while (iterator.hasNext() && recordList.size() < threshold) {
+          ConsumerRecord record = iterator.next();
+          recordList.add(record);
+        }
```

Comment:
The current implementation exits the loop when there are 3 empty polls. Having 3 empty *successive* polls, is probably a stronger guarantee. Will fix.